govcd/catalog.go

```diff
@@ -161,99 +165,148 @@ func (cat *Catalog) FindCatalogItem(catalogitem string) (CatalogItem, error) {
 	if err != nil {
-		return Task{}, err
+		return UploadTask{}, fmt.Errorf("%v. Unpacked files for checking are accessible in: "+tmpDir, err)
 	}
 
-	ovfUploadHref, err := getOvfUploadLink(vappTemplate)
+	catalogItemUploadURL, err := findCatalogItemUploadLink(cat)
 	if err != nil {
-		return Task{}, err
+		return UploadTask{}, err
 	}
 
-	filesAbsPaths, err := util.Unpack(ovaFileName)
+	vappTemplateUrl, err := createItemForUpload(cat.client, catalogItemUploadURL, itemName, description)
 	if err != nil {
-		return Task{}, err
+		return UploadTask{}, err
 	}
 
-	var ovfFileDesc Envelope
-	var tempPath string
+	vappTemplate, err := queryVappTemplate(cat.client, vappTemplateUrl, itemName)
+	if err != nil {
+		return UploadTask{}, err
+	}
 
-	for _, filePath := range filesAbsPaths {
-		if filepath.Ext(filePath) == ".ovf" {
-			ovfFileDesc, err = uploadOvfDescription(cat.client, filePath, ovfUploadHref)
-			tempPath, _ = filepath.Split(filePath)
-			if err != nil {
-				return Task{}, err
-			}
-			break
-		}
+	ovfUploadHref, err := getOvfUploadLink(vappTemplate)
+	if err != nil {
+		return UploadTask{}, err
 	}
 
-	vappTemplate, err = waitForTempUploadLinks(cat.client, vappTemplateUrl)
+	ovfFileDesc, err := uploadOvfDescription(cat.client, ovfFilePath, ovfUploadHref)
+	if err != nil {
+		return UploadTask{}, err
+	}
 
-	err = uploadFiles(cat.client, vappTemplate, &ovfFileDesc, tempPath, filesAbsPaths)
+	vappTemplate, err = waitForTempUploadLinks(cat.client, vappTemplateUrl, itemName)
 	if err != nil {
-		return Task{}, err
+		return UploadTask{}, err
+	}
+
+	var uploadProgress float64
+	callBack := func(bytesUploaded, totalSize int64) {
+		uploadProgress = (float64(bytesUploaded) / float64(totalSize)) * 100
 	}
 
+	//sending upload process to background, this allows no to lock and return task to client
+	go uploadFiles(cat.client, vappTemplate, &ovfFileDesc, tmpDir, filesAbsPaths, uploadPieceSize, callBack)
+
 	var task Task
 	for _, item := range vappTemplate.Tasks.Task {
 		task, err = createTaskForVcdImport(cat.client, item.HREF)
+		if err != nil {
+			return UploadTask{}, err
+		}
 	}
 
-	log.Printf("[TRACE] Upload finished and task for vcd import created. \n")
-	return task, nil
+	uploadTask := NewUploadTask(&task, &uploadProgress)
+
+	util.Logger.Printf("[TRACE] Upload finished and task for vcd import created. \n")
+
+	return *uploadTask, nil
 }
 
-func uploadFiles(client *Client, vappTemplate *types.VAppTemplate, ovfFileDesc *Envelope, tempPath string, filesAbsPaths []string) error {
+// Upload files for vCD created upload links. Different approach then vmdk file are
+// chunked (e.g. test.vmdk.000000000, test.vmdk.000000001 or test.vmdk). vmdk files are chunked if
+// in description file attribute ChunkSize is not zero.
+func uploadFiles(client *Client, vappTemplate *types.VAppTemplate, ovfFileDesc *Envelope, tempPath string, filesAbsPaths []string, uploadPieceSize int64, callBack func(bytesUpload, totalSize int64)) error {
```

Comment:
This function has too many positional arguments.
Please add documentation about the meaning of every argument. Consider doing the same for every function that has more than three args.